[MRG] Get multiple spaced events from single annotation

massich opened the pull request:
This PR adds a parameter to `events_from_annotation` so that if this parameter
None, events correspond to the annotation onsets (as was happening until nos).
But if not, events_from_annotations returns as many events as they fit within
the annotation duration spaced according to this new parameter (right now is
called `chunk_duraiton` but we should discuss it before merging).

This PR needs to go in before #5718 

---
Things that are not in this PR (it will be done in subsequent PRs):
- allow slicing of Annotations
- update the documentation of events and annotations

massich commented on mne/annotations.py:
```diff
@@ -741,49 +798,20 @@ def events_from_annotations(raw, event_id=None, regexp=None, use_rounding=True,
-
-    logger.info('Used Annotations descriptions: %s' %
-                (list(event_id_.keys()),))
+    for onset, duration, description in iterator:
```
I would like to add slicing of the annotations so that this can be:
`for onset, duration, description in raw.annotations[event_sel]:`

massich replied:
but this can be done afterwards

larsoner replied:
Yes a `__getitem__` that returns a copy and a `__iter__` method (or whatever the correct Py3k one is named) would be good at some point

massich commented:
@Slasnista you should be able to get the psg events just by doing the following:

```py
raw = mne.io.read_raw_edf(psg_fname, stim_channel=False)
annotations = mne.read_annotations(hyp_fname)
desc2int = {'Sleep stage W': 5,
￼             'Sleep stage 1': 3,
￼             'Sleep stage 2': 2,
￼             'Sleep stage 3': 1,
￼             'Sleep stage 4': 1,
￼             'Sleep stage R': 4}
events, event_id = mne.events_from_annotations(raw, event_id=desc2int, chunk_duration=30)
```

massich commented:
I've reverted the last commits where I refactored the code to be simpler because it was messing up with EDF and to fix it would complicate the PR. I'll get back to it once EDF #5741 is in and `find_edf_events` is out.

agramfort commented:
@massich you have some pep8 errors.

massich commented on mne/annotations.py:
```diff
@@ -700,16 +700,60 @@ def _ensure_annotation_object(obj):
 @verbose
-def events_from_annotations(raw, event_id=None, regexp=None, use_rounding=True,
-                            verbose=None):
+def events_from_annotations(raw, chunk_duration=None, event_id=None,
```
True. I put it there 'cos it seems to be more logical, but I didn't think about killing API

agramfort commented on mne/tests/test_annotations.py:
```diff
@@ -185,6 +185,33 @@ def test_crop():
+    If chunk_duration parameter in events_from_annotations is None, events
+    correspond to the annotation onsets. If not, events_from_annotations
+    returns as many events as they fit within the annotation duration spaced
+    according to `chunk_duraiton`.
```
chunk_duraiton  -> chunk_duration

massich commented on mne/annotations.py:
```diff
@@ -700,16 +700,60 @@ def _ensure_annotation_object(obj):
+    """Gets a collection of descriptions and returns index of selected.
+    """
+    # Filter out the annotations that do not match regexp
+    regexp_comp = re.compile('.*' if regexp is None else regexp)
```
I don't get what you mean. If I change the function's signature from `regexp=None` to `regexp='.*'` then I'll have to make the check outside and only add the regex parameter if it is not none; and we rather prefer the inner function doing that, wouldn't we?

larsoner replied:
With @agramfort's suggestion you never need to check `if regexp is None` anymore. What check will you still need to do?

massich replied:
To avoid this situation:
```
def events_from_annotations(raw, event_id=None, regexp=None, use_rounding=True,
                            verbose=None):
    ...
    _select_annotations_based_on_description(descriptions, regexp=regexp)
    ...

def _select_annotations_based_on_description(descriptions, event_id=None,
                                             regexp='.*'):
     assert regexp is None
```

Otherwise, I've to deprecate the `regexp=None`

larsoner replied:
Instead of `assert regexp is None` inside `_select_annotataions...`, did you mean the opposite, `assert regexp is not None`?

I agree with @agramfort that `regexp='.*'` is a nicer default than `None`. And it simplifies the inner function to know that `regexp` is a string. So I would do:
```
def events_from_annotations(..., regexp='.*', ...)
    """...
    regexp : str
        ...
    """
    # regexp None support for 0.17 backward compat, not exposed to users
    regexp = '.*' if regexp is None else regexp
    _validate_type(regexp, str, 'regexp')

...
def _select_annotations_based_on_description(descriptions, event_id, regexp):
     assert isinstance(regexp, str)
```

massich replied:
The assert was just to illustrate. If we don't have to deprecate the old behavior with a deprecation cycle and everything. Then I agree.

larsoner replied:
We shouldn't need to since `'.*'` will match everything just like `None` would have.

The only difference is whether an error is raised if no events are found. I'm not sure we should have ever differentiated based on whether `regexp` was passed by the user or not. The most consistent thing to do would be to have an `on_missing` argument like we do for `Epochs`. I think it's okay to change this behavior without deprecation since it's going to be a corner case.

agramfort commented on mne/annotations.py:
```diff
@@ -701,9 +701,48 @@ def _ensure_annotation_object(obj):
+                                             regexp=None):
+    """Get a collection of descriptions and returns index of selected."""
+
+    regexp_comp = re.compile(regexp)
```
this is weird. You have None as default but you cannot call re.compile with None.

agramfort commented on mne/annotations.py:
```diff
@@ -741,47 +786,38 @@ def events_from_annotations(raw, event_id=None, regexp=None, use_rounding=True,
-            else:
-                dropped.append(desc)
+    # regexp None support for 0.17 backward compat, not exposed to users
+    regexp = '.*' if regexp is None else regexp
```
to remove now that regexp cannot be None

agramfort commented:
@massich you need to rebase

massich commented:
@larsoner can you merge?

massich commented on mne/annotations.py:
```diff
@@ -701,9 +701,47 @@ def _ensure_annotation_object(obj):
+    event_sel = [ii for ii, kk in enumerate(descriptions)
+                 if kk in event_id_]
+
+    if len(event_sel) == 0 and regexp is not None:
```
actually needs to compare with the default regex `'.*'`, and it's not even covered. I'll add this `on_missing` thing. Do you know how we test the logging from the top of your head?

larsoner replied:
I wouldn't bother testing if the `logger.info` is actually emitted, it's probably overkill. If you test that by default it errors, and that with `warning` it warns, and that with `info` (or whatever we call it) it does neither (which can be done just by a smoke test since an un-caught error or warning would be a test error) it should be fine.

But if you really want to go for it you can use `mne.utils.catch_logging`

massich replied:
since they are both the same, now we could factor them and have it as a `good first issue` make it into a function and cover it fully so that we don't have to go in the `'error'` etc.

massich commented:
After a talk IRL with @agramfort we decided to set back regexp to `None` instead of `'.*'`. And revoked the possibility of warning/raising to avoid the parameter at the API. 

I see the value of `on_missing` (that was not the case for the `None`) but I agree that havning `on_missing`, `verbose` .. it keeps polluting the API. But if someone things that would rather have warning instead of error I don't mind revert 71623e8 ( and add a **good first issue** for someone to factor the code).

larsoner commented on mne/tests/test_annotations.py:
```diff
@@ -530,7 +551,7 @@ def test_events_from_annot_in_raw_objects():
 
     with pytest.raises(ValueError, match='not find any of the events'):
-        events_from_annotations(raw, regexp='not_there')
+        events_from_annotations(raw, regexp='not_there', on_missing='error')
```
`on_missing` is still here

larsoner commented on mne/annotations.py:
```diff
@@ -701,9 +701,47 @@ def _ensure_annotation_object(obj):
 
 
+def _select_annotations_based_on_description(descriptions, event_id=None,
+                                             regexp=None, on_missing='ignore'):
```
on_missing is still here

massich commented:
Did I forgot to take them out from the signature? I'm AFK. Feel free to
take them out and merge if you want.

Other wise if you think that we should revert the last commit, feel free to
do so.

If not I'll remove the on_missing once I get back

On Wed, Dec 19, 2018, 19:20 Eric Larson <notifications@github.com> wrote:

> *@larsoner* commented on this pull request.
> ------------------------------
>
> In mne/tests/test_annotations.py
> <https://github.com/mne-tools/mne-python/pull/5795#discussion_r243023592>:
>
> > @@ -530,7 +551,7 @@ def test_events_from_annot_in_raw_objects():
>      assert_array_equal(expected_events5, events5)
>
>      with pytest.raises(ValueError, match='not find any of the events'):
> -        events_from_annotations(raw, regexp='not_there')
> +        events_from_annotations(raw, regexp='not_there', on_missing='error')
>
> on_missing is still here
> ------------------------------
>
> In mne/annotations.py
> <https://github.com/mne-tools/mne-python/pull/5795#discussion_r243023660>:
>
> > @@ -701,9 +701,47 @@ def _ensure_annotation_object(obj):
>                           'mne.Annotations. Got %s.' % obj)
>
>
> +def _select_annotations_based_on_description(descriptions, event_id=None,
> +                                             regexp=None, on_missing='ignore'):
>
> on_missing is still here
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/5795#pullrequestreview-186683096>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGt-44bbw-0QdHZPHibxayU0sFievb2Oks5u6oNSgaJpZM4ZV5Fb>
> .
>
